docs/proposals/security-context-constraints.md

```diff
@@ -0,0 +1,320 @@
+is up to the cluster administrator to ensure that a security context constraint object exists that
+references the group.
+
+If an administrator would like to provide a user with a changed set of security context permissions
+they may do the following:
+
+1.  Create a new security context constraints object and add a reference to the user or a group
+that the user belongs to.
+1.  Add the user (or group) to an existing security context constraints object with the proper
+elevated privileges.
+
+## Admission
+
+Admission control using an authorizer allows the ability to control the creation of resources
+based on capabilities granted to a user.  In terms of the security context constraints it means
+that an admission controller may inspect the user info made available in the context to retrieve
+and appropriate set of security context constraints for validation.
+
+The appropriate set of security context constraints is defined as all of the security context
+constraints available that have reference to the user or groups that the user belongs to.
+
+Admission will use the SecurityContextConstraintsAllocator to ensure that any requests for a
+specific security context constraint setting or to generate settings using the following approach:
+
+**Pod with referenced service account**
+
+1.  First, ensure that the user has the ability to use the security context constraints of the
+service account.  This can be accomplished by finding the intersection of allowed constraints between
+the user and the service account.  If none exists then the user does not have permissions to
+run as the service account.
+2.  Ask the SecurityContextConstraintsAllocator to create the security context.  The allocator
+will not overwrite fields already set in the container.
+3.  Validate that the generated SecurityContext falls within the boundaries of the security context
+constraints and accept or reject the pod.
+
+**Pods with no service account and pods whose service account does not match the requested privileges**
+
+1.  Retrieve all security context constraints available for use by the user.
+2.  Loop through the constraints to ensure that any requests on the pod fall within an allowed
+security context constraint.  This constraint is what will be used in subsequent steps.  If no
+acceptable constraint is found then reject the pod.
+2.  Ask the SecurityContextConstraintsAllocator to create the security context.  The allocator
+will not overwrite fields already set in the container.
+3.  Validate that the generated SecurityContext falls within the boundaries of the security context
+constraints and accept or reject the pod.
+
+Note on validation:  Since a user may have more than one security context constraint that they
+are allowed to use validation should take place in two steps.
+
+1.  Soft Validation: Ensure that any requests on the pod fall within the constraint.  If a field like RunAsUser
```

Comment:
TODO: remove this, no need for multi-step validation.  The `Generate` call should not change fields that are already set and the `Validate` call should just fail if the generated context is not within the bounds of the SCC.